### ISSUE TITLE
Implement a timeout when using unix sockets with weave to avoid the edge-case of always blocking on an already existing socket.

### ISSUE THREAD

**Issue by ruseinov** [iov-one/weave#273]
This is really hard to debug as any crashes produced by the first run may very well be hidden by hanging up on a second run, especially if tendermint is poking at that same socket.

*TODO* Design, review exactly what changes needed, what easiest... remember abci server is in tendermint code based

**Comment by ethanfrey** [comment]
This is a proposed solution, with a change to the abci server (in tendermint).

I would like to see the clear problem well defined, so we can consider multiple solutions. I think the problem is: `if bnsd binary crashes, the original issue is impossible to find in logs due to restarts, make this easier to debug`

**Comment by ethanfrey** [comment]
Before stating this task, please do some research on possible solutions, such that the k8 logs show the root cause of failures, and ensure the dev team agrees with the design.

Ideally the proposal is before or at the beginning of the sprint, to leave some days for feedback before starting this code

**Comment by alpe** [comment]
We should see the logs in gcp stackdriver but that is usually not the first place to look for them but `kubectl logs`. 
Another approach would be to rethink our current tendermint/abci app separation. It would be easier from OPS perspective to run only 1 process as Gaiad does (https://github.com/cosmos/cosmos-sdk/blob/master/server/start.go#L124)

**Comment by ethanfrey** [comment]
The downside with one app is harder to update and configure tendermint. We have to update all the p2p flags and other options in the Bnsd binary.

Not critical, but just some work to do. 

I do agree ops is easier. 

If you all like that, then I can make a clear ticket on how to integrate, as I wrote most of the tendermint and cosmos Sdk cli code, and most still seems to be in use 18 months later

**Comment by ruseinov** [comment]
In my opinion running several processes (apps) as one app is actually a maintenance nightmare.
I also don't like running more than one concern in a docker container, while that might seem easier at a first glance, separated concerns have their benefits and it's easier to spot what went wrong and where.

But that's me

**Comment by ruseinov** [comment]
I'm actually pretty happy with the fact that weave is separated from tendermint, not just from ops perspective, but in general.
Also because should we ever want to support a different consensus/whatnot - it's going to be more natural than a coupled solution.

**Comment by ruseinov** [comment]
One plus side though: for users it would be easier to run. That's 100% true

**Comment by ethanfrey** [comment]
When run in process, we just swap out the Unix socket or TCP socket for a channel.

A bit harder to monitor where memory goes etc, but not exactly two apps run as one. We more run it as the one app split into two variant. 

That said, I somewhat prefer the split variant, only the ops overhead sucks sometimes

**Comment by ruseinov** [comment]
Well, it’s not the overhead that will save us that much headache tbh, as we have already dealt with the setup.

So unless @alpe has a strong preference - I would prefer to keep it as is. Otherwise I’m not gonna push for either. 

To me it’s cleaner to have it split.

**Comment by alpe** [comment]
I prefer the single app setup to be honest. Both apps feel tightly coupled anyway although the ABCI interface exists. It would be  easier to start and easier to explain to somebody else IMHO. 

But I don't want to push my preferences. Current setup is ok, too but this issue would not exist with a single app. That's why I brought it up.

**Comment by ruseinov** [comment]
@alpe I'm with you on ease of use.
I have a suspicion though that depending on tendermint 100% is going to make this thing a monster.

When I was working with gaia I didn't actually like that it contained code, introspecting which you'd have to dig in to 1000s of lines of tendermint code, it's overwhelming even if you know what it happening. 

One more valid argument for a merge of the two apps is that actual nodes on production nets are not supposed to be run in Docker anyway.

And the last one: I think this issue is actually low priority as it helps in some cases when deploying testnets, but otherwise it does not matter that much.

**Comment by alpe** [comment]
fyi: I had run into Irisnet recently and they do single app, too: https://github.com/irisnet/irishub/blob/master/cmd/iris/main.go#L113

**Comment by ethanfrey** [comment]
@alpe does that comment mean you think we go single-binary as well?

Irisnet is following cosmos-sdk, so of course they use the same process.

I think two apps is cleaner in a way, also will allow us to make the abci app with minimum dependencies if we extract tendermint and then easier to audit. But if it is too much overhead for ops, let's go with one app.

**Comment by alpe** [comment]
Sorry for the confusion. 2 apps are not too much overhead we can not
handle.
For UX, doc 1 app is better IMHO. I am waiting for binance to release
theirs to post 1 app is the standard. :-)

Ethan Frey <notifications@github.com> schrieb am Mo., 4. März 2019, 16:55:

**Comment by ethanfrey** [comment]
binance and irisnet both just follow cosmos-sdk with no original architecture afaik.
They only validate that people forking cosmos-sdk and following their architecture use 1 app.
It does not indicate a deep reflection convinced them this is better.

I wonder how well they support all the extra tendermint flags, like `--p2p.max_peers=...` or if this must be done in the config file?

**Comment by husio** [comment]
This issue is on top of `Refined Backlog` column. I believe it is not yet ready to be worked on as there is no clear description of the problem or a functionality that is to be implemented.
Can this story go back to the first column?

**Comment by ethanfrey** [comment]
Good point @husio Let's put it back with needs design

**Comment by ruseinov** [comment]
closed in favour https://github.com/iov-one/weave/issues/851
